File: src/bosh_openstack_cpi/lib/cloud/openstack/volume_configurator.rb
module Bosh::OpenStackCloud
  class VolumeConfigurator
    include Helpers

    def initialize(logger)
      @logger = logger
    end

    def select_boot_volume_size(flavor, resource_pool)
      if resource_pool['root_disk'].nil?
        if flavor.disk == 0
          cloud_error("Flavor '#{flavor.name}' has a root disk size of 0. Either pick a different flavor or define root_disk.size in your VM cloud_properties")
        end

        flavor.disk
      else
        root_disk_size = resource_pool['root_disk']['size']
        raise ArgumentError, 'Minimum root_disk size is 1 GiB' if root_disk_size.nil? || root_disk_size < 1
        @logger.debug("Using root_disk of size '#{root_disk_size}', instead of flavor.disk")

        root_disk_size
      end
    end

    def boot_from_volume?(boot_from_volume, resource_pool)
      return boot_from_volume if resource_pool['boot_from_volume'].nil?

      resource_pool['boot_from_volume']
    end
  end
end

File: src/openstack_cpi_golang/cpi/compute/volume_configurator.go
package compute

import (
	"fmt"

	"github.com/cloudfoundry/bosh-openstack-cpi-release/src/openstack_cpi_golang/cpi/config"
	"github.com/cloudfoundry/bosh-openstack-cpi-release/src/openstack_cpi_golang/cpi/properties"
	"github.com/gophercloud/gophercloud/openstack/compute/v2/extensions/bootfromvolume"
	"github.com/gophercloud/gophercloud/openstack/compute/v2/flavors"
)

//counterfeiter:generate . VolumeConfigurator
type VolumeConfigurator interface {
	ConfigureVolumes(
		imageID string,
		openstackConfig config.OpenstackConfig,
		cloudProperties properties.CreateVM,
		flavor flavors.Flavor,
	) ([]bootfromvolume.BlockDevice, error)
}

type volumeConfigurator struct{}

func NewVolumeConfigurator() volumeConfigurator {
	return volumeConfigurator{}
}
func (v volumeConfigurator) ConfigureVolumes(imageID string, openstackConfig config.OpenstackConfig, cloudProperties properties.CreateVM, flavor flavors.Flavor) ([]bootfromvolume.BlockDevice, error) {
	bootVolumeSize, err := v.select_boot_volume_size(flavor, cloudProperties)
	if err != nil {
		return []bootfromvolume.BlockDevice{}, fmt.Errorf("failed to get volume size: %w", err)
	}

	if !v.bootFromVolume(openstackConfig, cloudProperties) {
		return []bootfromvolume.BlockDevice{}, nil
	}

	return []bootfromvolume.BlockDevice{{
		UUID:                imageID,
		SourceType:          bootfromvolume.SourceImage,
		DestinationType:     bootfromvolume.DestinationVolume,
		VolumeSize:          bootVolumeSize,
		BootIndex:           0,
		DeleteOnTermination: true,
	}}, nil
}

func (v volumeConfigurator) bootFromVolume(openstackConfig config.OpenstackConfig, cloudProperties properties.CreateVM) bool {
	if cloudProperties.BootFromVolume == nil {
		return openstackConfig.BootFromVolume
	}

	return *cloudProperties.BootFromVolume
}

func (v volumeConfigurator) select_boot_volume_size(flavor flavors.Flavor, cloudProperties properties.CreateVM) (int, error) {
	rootDiskSize := cloudProperties.RootDisk.Size
	if rootDiskSize == 0 {
		if flavor.Disk == 0 {
			return 0, fmt.Errorf("flavor '%s' has a root disk size of 0. Either pick a different flavor or define root_disk.size in your VM cloud_properties", flavor.ID)
		}
		return flavor.Disk, nil
	}

	return rootDiskSize, nil
}
